core/src/test/scala/org/apache/spark/sql/delta/StatsBasedDataSkippingSuite.scala

```diff
@@ -0,0 +1,267 @@
+      df2.write.format("delta").mode(SaveMode.Append).save(testTablePath)
+      df2.write.format("delta").mode(SaveMode.Append).saveAsTable(mixedStatsTableName)
+    }
+  }
+
+  test("Select Count: basic") {
+    checkResultsAndOptimizedPlan(
+      s"SELECT COUNT(*) FROM $testTableName",
+      Seq(Row(totalRows)),
+      "LocalRelation [none#0L]")
+  }
+
+  test("Select Count: column alias") {
+    checkResultsAndOptimizedPlan(
+      s"SELECT COUNT(*) as MyColumn FROM $testTableName",
+      Seq(Row(totalRows)),
+      "LocalRelation [none#0L]")
+  }
+
+  test("Select Count: table alias") {
+    checkResultsAndOptimizedPlan(
+      s"SELECT COUNT(*) FROM $testTableName MyTable",
+      Seq(Row(totalRows)),
+      "LocalRelation [none#0L]")
+  }
+
+  test("Select Count: time travel") {
+    checkResultsAndOptimizedPlan(s"SELECT COUNT(*) FROM $testTableName VERSION AS OF 0",
+      Seq(Row(3L)),
+      "LocalRelation [none#0L]")
+
+    checkResultsAndOptimizedPlan(s"SELECT COUNT(*) FROM $testTableName VERSION AS OF 1",
+      Seq(Row(2L)),
+      "LocalRelation [none#0L]")
+
+    checkResultsAndOptimizedPlan(s"SELECT COUNT(*) FROM $testTableName VERSION AS OF 2",
+      Seq(Row(totalRows)),
+      "LocalRelation [none#0L]")
+  }
+
+  test("Select Count: external") {
+    checkResultsAndOptimizedPlan(
+      s"SELECT COUNT(*) FROM delta.`$testTablePath`",
+      Seq(Row(totalRows)),
+      "LocalRelation [none#0L]")
+  }
+
+  test("Select Count: sub-query") {
+    checkResultsAndOptimizedPlan(
+      s"SELECT (SELECT COUNT(*) FROM $testTableName)",
+      Seq(Row(totalRows)),
+      "Project [scalar-subquery#0 [] AS #0L]\n:  +- LocalRelation [none#0L]\n+- OneRowRelation")
+  }
+
+  test("Select Count: as sub-query filter") {
+    checkResultsAndOptimizedPlan(
+      s"SELECT 'ABC' WHERE (SELECT COUNT(*) FROM $testTableName) = $totalRows",
+      Seq(Row("ABC")),
+      "Project [ABC AS #0]\n+- Filter (scalar-subquery#0 [] = " +
+        totalRows + ")\n   :  +- LocalRelation [none#0L]\n   +- OneRowRelation")
+  }
+
+  test("Select Count: limit") {
+    // Limit doesn't affect COUNT results
+    checkResultsAndOptimizedPlan(
+      s"SELECT COUNT(*) FROM $testTableName LIMIT 3",
+      Seq(Row(totalRows)),
+      "LocalRelation [none#0L]")
+  }
+
+  // Tests to validate the optimizer won't use missing or partial stats
+  test("Select Count: missing stats") {
+    checkSameQueryPlanAndResults(
+      s"SELECT COUNT(*) FROM $mixedStatsTableName",
+      Seq(Row(totalRows)))
+
+    checkSameQueryPlanAndResults(
+      s"SELECT COUNT(*) FROM $noStatsTableName",
+      Seq(Row(totalRows)))
+  }
+
+
+  // Tests to validate the optimizer won't incorrectly change queries it can't correctly handle
+
+  test("Select Count: multiple aggregations") {
+    checkSameQueryPlanAndResults(
+      s"SELECT COUNT(*) AS MyCount, MAX(id) FROM $testTableName",
+      Seq(Row(totalRows, 10L)))
+  }
+
+  test("Select Count: group by") {

```

Comment:
A frequent use case I have is grouping by a partitioning column, is this tested anywhere? If not, would it be possible to add a test case?

Reply:
I can add the test. But both GROUP BY DataColumn or PartitionColumn are not supported by this optimization. The test here is validating that nothing changes where query has GROUP BY.
I guess it is theoretically possible to count using the stats when GROUP BY PartitionColumn or PartitionColumn = X, but I'm intentionally not implementing it here, I'd like to first validate the overall design, before optimizing more complex queries.

Reply:
sure, that makes total sense. thank you for the explanation